docs/proposals/affinity.md

```diff
@@ -0,0 +1,341 @@
+	// Satisfying more expressions is better than satisfying fewer expressions.
+	SoftAffinityExpressions []AffinityTerm `json:"softAffinityExpression,omitempty"`
+	// Satisfying more expressions is better than satisfying fewer expressions.
+	SoftAntiAffinityExpressions []AffinityTerm `json:"softAntiAffinityExpression,omitempty"`
+}
+
+type AffinityTerm struct {
+	PodSelector *PodSelector `json:"podSelector,omitempty"` // empty PodSelector matches all pods
+	TopologyKey string `json:"topologyKey,omitempty"`
+}
+```
+
+One caveat: As noted above, the normal semantics of empty `PodSelector` is to match all pods.
+However, `PodSelector` is scoped to namespace, therefore the normal semantics do not allow
+us to match all pods globally. We need to special-case interpret empty `PodSelector` in this
+context to mean "all pods globally."
+
+To explain how this API works, let's say that the `PodSpec` of a pod `P` has an `AffinitySelector`
+that is configured as follows:
+
+```go
+AffinitySelector {
+	HardAffinityExpressions: {{PodSelector: P1, TopologyKey: "node"}},
+	HardAntiAffinityExpressions: {{PodSelector: P2, TopologyKey: "rack"}},
+	SoftAffinityExpressions: {{PodSelector: P3, TopologyKey: "zone"}},
+	SoftAntiAffinityExpressions: {{PodSelector: P4, TopologyKey: "power"}}
+}
+```
+
+Then when scheduling pod P, the scheduler
+* Can only schedule P onto nodes that are running pods that satisfy `P1`. (Assumes all nodes have a label with key "node" and value specifying their node name.)
+* Cannot schedule P onto any racks that are running pods that satisfy `P2`. (Assumes all nodes have a label with key "rack" and value specifying their rack name.)
+* Should try to schedule P onto zones that are running pods that satisfy `P3`. (Assumes all nodes have a label with key "zone" and value specifying their zone.)
+* Should try not to schedule P onto any power domains that are running pods that satisfy `P4`. (Assumes all nodes have a label with key "power" and value specifying their power domain.)
+
+When `HardAffinityExpressions` has multiple elements, the requirements are ANDed. Likewise for `HardAntiAffinityExpressions`.
+For `SoftAffinityExpressions` and `SoftAntiAffinityExpressions`, nodes that satisfy more of the rules are preferred over those that satisfy fewer.
+
+## A comment on symmetry
+
+One thing that makes affinity and anti-affinity tricky is symmetry.
+
+Imagine a cluster that is running pods from two services, S1 and S2. Imgine that the pods of S1 have a hard anti-affinity rule
+"do not run me on nodes that are running pods from S2." It is not sufficient just to check that there are no S2 pods on a node when
+you are scheduling a S1 pod. You also need to ensure that there are no S1 pods on a node when you are scheduling a S2 pod,
+*even though the S2 pod does not have any anti-affinity rules*. Otherwise if an S1 pod schedules before an S2 pod, the S1
+pod's hard anti-affinity rule can be violated by a later-arriving S2 pod.
```

Comment:
I think we have some inconsistency around this argument that rules need to remain satisfied over time.  In this example, why do we try to keep S1's anti-affinity constraint satisfied after S1 has been scheduled (e.g. when scheduling S2), but we don't try to keep S1's constraints satisfied after it's scheduled in other cases (see "We assume pods that become infeasible after they are bound to a node may be allowed to continue running") below.  For example, if S1 has a hard affinity to run only on nodes running service S2, we could schedule S1 to fulfill that constraint, and then promptly terminate all pods in S2.  Should we then also terminate all pods in S1?  I gather that you're suggesting that we do not.  But then what's the use of calling the affinity "hard" in the first place, if we can violate it at some times (e.g. on termination of B) but not others (e.g. on scheduling or rescheduling of A.  This often leads to all sorts of very confusing system behavior (for example, all replicas of service A are running just fine after termination of service B, but then some of service A's pods crash, but cannot be rescheduled because B's pods no longer exist).  It seems we have two choices:

1. Just call all affinities soft (and stop fooling ourselves that they're hard :-) when they're not, as per above example)
2. Come up with a good sane explanation of what exactly "hard" means, rather than defer that to being "out of scope".

 

It seems very confusing to me that in some cases we try to keep constraints satisfied over time, but in other cases not.  I realize that you deemed this to be beyond the scope of this document (in the comment below), but until we have a good answer to that apparent inconsistency, the whole notion of HardAffinity is potentially of little or no value.